### ISSUE TITLE
Allow specifying creds and subscription-id

### ISSUE THREAD

**Issue by onionhammer** [Azure/login#282]
Relates to #276 since it was closed without a valid reason and is now being ignored.

It appears that you can only specify one or the other, but it would be useful to be able to override settings like `subscription-id` in `creds` if present for cases where you have one credential with access to multiple subscriptions on different environments

This code will override subscriptionId from input if creds are provided:

```ts
let creds = core.getInput('creds', { required: false });
let secrets = creds ? new SecretParser(creds, FormatType.JSON) : null;
let environment = core.getInput("environment").toLowerCase();
const enableAzPSSession = core.getInput('enable-AzPSSession').toLowerCase() === "true";
const allowNoSubscriptionsLogin = core.getInput('allow-no-subscriptions').toLowerCase() === "true";

//Check for the credentials in individual parameters in the workflow.
var servicePrincipalId = core.getInput('client-id', { required: false });
var servicePrincipalKey = null;
var tenantId = core.getInput('tenant-id', { required: false });
var subscriptionId = core.getInput('subscription-id', { required: false });
var resourceManagerEndpointUrl = "https://management.azure.com/";
var enableOIDC = true;
var federatedToken = null;

// If any of the individual credentials (clent_id, tenat_id, subscription_id) is present.
if (servicePrincipalId || tenantId || subscriptionId) {

    //If few of the individual credentials (clent_id, tenat_id, subscription_id) are missing in action inputs.
    if (!(servicePrincipalId && tenantId && (subscriptionId || allowNoSubscriptionsLogin)))
        throw new Error("Few credentials are missing. ClientId, tenantId are mandatory. SubscriptionId is also mandatory if allow-no-subscriptions is not set.");
}
else {
    if (creds) {
        core.debug('using creds JSON...');
        enableOIDC = false;
        servicePrincipalId = secrets.getSecret("$.clientId", true);
        servicePrincipalKey = secrets.getSecret("$.clientSecret", true);
        tenantId = secrets.getSecret("$.tenantId", true);
        subscriptionId = secrets.getSecret("$.subscriptionId", true);
        resourceManagerEndpointUrl = secrets.getSecret("$.resourceManagerEndpointUrl", false);
    }
    else {
        throw new Error("Credentials are not passed for Login action.");
    }
}
```

Here are the relevant parts:

```ts
let creds = core.getInput('creds', { required: false });
let secrets = creds ? new SecretParser(creds, FormatType.JSON) : null;
// snip ...
var subscriptionId = core.getInput('subscription-id', { required: false });
// snip ...

// IF subscription id set, 'creds' object is completely ignored
if (servicePrincipalId || tenantId || subscriptionId) {

    //If few of the individual credentials (clent_id, tenat_id, subscription_id) are missing in action inputs.
    if (!(servicePrincipalId && tenantId && (subscriptionId || allowNoSubscriptionsLogin)))
        throw new Error("Few credentials are missing. ClientId, tenantId are mandatory. SubscriptionId is also mandatory if allow-no-subscriptions is not set.");
}
else {
    if (creds) {
        core.debug('using creds JSON...');
        enableOIDC = false;
        servicePrincipalId = secrets.getSecret("$.clientId", true);
        servicePrincipalKey = secrets.getSecret("$.clientSecret", true);
        tenantId = secrets.getSecret("$.tenantId", true);
        // HERE the subscription id is overridden by the value in the 'creds' object
        subscriptionId = secrets.getSecret("$.subscriptionId", true);
        resourceManagerEndpointUrl = secrets.getSecret("$.resourceManagerEndpointUrl", false);
    }
    else {
        throw new Error("Credentials are not passed for Login action.");
    }
}
```

creds appears to override every single other value, not the other way around.

**Comment by BALAGA-GAYATRI** [comment]
Hi @onionhammer , 
As explained here - https://github.com/Azure/login/issues/276#issuecomment-1340370067
if you look at the if else condition here, first we are checking for the presence of any of the individual credentials - 
`if (servicePrincipalId || tenantId || subscriptionId` and if they are present we are going with OIDC approach and wouldn't even enter into the non-oidc condition. And if the individual credentials are not given, we are looking for the `creds` json object and will go through non-oidc approach. In case both creds and individual credentials are provided, individual credentials will have the presence over the creds object. This is designed to encourage customers to go with the OIDC login mechanism and also to address this - 
> To help ensure that GitHub redacts your secret in logs, avoid using structured data as the values of secrets. For example, avoid creating secrets that contain JSON or encoded Git blobs.


https://docs.github.com/en/actions/security-guides/encrypted-secrets#naming-your-secrets

Hope this answers your question.

**Comment by onionhammer** [comment]
@BALAGA-GAYATRI 

>  In case both creds and individual credentials are provided, individual credentials will have the presence over the creds object. 

This needs to be tested by you if you don't believe me, but I would assert that the credentials in the `creds` object have precedence over the individual credentials, not vice versa as you suggest.

**Comment by BALAGA-GAYATRI** [comment]
Can you please share your workflow and debug log? That will help us identifying the problem better.

**Comment by onionhammer** [comment]
@BALAGA-GAYATRI Here is an example snippet of a workflow;

```yml
uses: azure/login@v1
        with:
          creds: ${{ secrets.AZURE_CREDENTIALS }}
          subscription-id: ${{ secrets.SUBSCRIPTION_PROD }}
      - run: az account show
```

This will fail because of this code

```typescript
if (servicePrincipalId || tenantId || subscriptionId) {

    //If few of the individual credentials (clent_id, tenat_id, subscription_id) are missing in action inputs.
    if (!(servicePrincipalId && tenantId && (subscriptionId || allowNoSubscriptionsLogin)))
        throw new Error("Few credentials are missing. ClientId, tenantId are mandatory. SubscriptionId is also mandatory if allow-no-subscriptions is not set.");
}
```

What my suggestion is is that you should instead allow individually specified items to override the `creds` object.

**Comment by BALAGA-GAYATRI** [comment]
Got your point. This is by the design choice. 
As of now, the only way for customers to choose whether they want to go with the OIDC / Non-OIDC is by the input parameters only. If the user provides individual credentials then OIDC login is triggered. Otherwise, non-OIDC login will happen. If the user provides only one individual credential (with or without providing creds JSON object), we take it as input for OIDC login and prompt the user to provide all of them in case anything is missing. If we allow the suggestion in the issue, it would be confusing for the customers also to opt for OIDC/Non-OIDC login. 
Hope this helps with your question.

**Comment by onionhammer** [comment]
I don't think it would be 'confusing', it just needs to be documented. The logic currently is very cryptic and hard to follow; this proposal would add a lot of flexibility.

**Comment by BALAGA-GAYATRI** [comment]
@onionhammer We shall definitely keep a note of this feedback and will try to implement but as of now changing the input formats for individual login mechanisms is not on the roadmap. Hence closing this issue.

**Comment by onionhammer** [comment]
The input format does not have to be changed, just the result of specifying yields unexpected behavior.
